cmd/ebpf/xdp/n3n6_entrypoint.c

```diff
@@ -39,34 +39,81 @@
 #include "xdp/utils/routing.h"
 #include "xdp/utils/icmp.h"
 
-
-
 #define DEFAULT_XDP_ACTION XDP_PASS
 
 static __always_inline enum xdp_action send_to_gtp_tunnel(struct packet_context *ctx, int srcip, int dstip, __u8 tos, int teid) {
     if (-1 == add_gtp_over_ip4_headers(ctx, srcip, dstip, tos, teid))
         return XDP_ABORTED;
-  
     upf_printk("upf: send gtp pdu %pI4 -> %pI4", &ctx->ip4->saddr, &ctx->ip4->daddr);
     increment_counter(ctx->n3_n6_counter, tx_n3);
     return route_ipv4(ctx->xdp_ctx, ctx->eth, ctx->ip4);
 }
 
-static __always_inline enum xdp_action handle_n6_packet_ipv4(struct packet_context *ctx) {
+static __always_inline __u8 match_sdf_filter_ipv4(const struct iphdr *ip4, struct sdf_filter *sdf) {
+    __u8 packet_protocol = ip4->protocol;
+    __u32 packet_src_ip = ip4->saddr;
+    __u32 packet_dst_ip = ip4->daddr;
+
+
+    if (sdf->protocol != packet_protocol ||
+        (packet_src_ip & sdf->src_addr.mask) != sdf->src_addr.ip || 
+        (packet_dst_ip & sdf->dst_addr.mask) != sdf->dst_addr.ip) {
+        return 0;
+    }
+
+    upf_printk("Packet with source ip:%pI4 and destination ip:%pI4 matches SDF filter", &ip4->saddr, &ip4->daddr);
+
+    return 1;
+}
+
+static __always_inline __u8 match_sdf_filter_ipv6(const struct ipv6hdr *ip6, struct sdf_filter *sdf) {
+    __u8 packet_protocol = ip6->nexthdr;
+    struct in6_addr packet_src_ip = ip6->saddr;
+    struct in6_addr packet_dst_ip = ip6->daddr;
+
+    __uint128_t packet_src_ip_128 = *((__uint128_t*)packet_src_ip.s6_addr);
+    __uint128_t packet_dst_ip_128 = *((__uint128_t*)packet_dst_ip.s6_addr);
+
+    if (sdf->protocol != packet_protocol ||
+        (packet_src_ip_128 & sdf->src_addr.mask) != sdf->src_addr.ip || 
+        (packet_dst_ip_128 & sdf->dst_addr.mask) != sdf->dst_addr.ip) {
+        return 0;
+    }
+
+    upf_printk("Packet with source ip:%pI6 and destination ip:%pI6 matches SDF filter", &ip6->saddr, &ip6->daddr);
+
+    return 1;
+}
+
+static __always_inline __u16 handle_n6_packet_ipv4(struct packet_context *ctx) {
     const struct iphdr *ip4 = ctx->ip4;
     struct pdr_info *pdr = bpf_map_lookup_elem(&pdr_map_downlink_ip4, &ip4->daddr);
     if (!pdr) {
         upf_printk("upf: no downlink session for ip:%pI4", &ip4->daddr);
         return DEFAULT_XDP_ACTION;
     }
 
-    struct far_info *far = bpf_map_lookup_elem(&far_map, &pdr->far_id);
+    struct sdf_filter *sdf = &pdr->sdf_rules.sdf_filter;
+    __u32 far_id = pdr->far_id;
+    __u32 qer_id = pdr->qer_id;
+    __u8 outer_header_removal = pdr->outer_header_removal;
+    if(sdf) {
```

Comment:
Always true here, actually. See the pointer initialization. You need some criterion for a SDF filter presence detection.